File: Data_Structures/Trees/binary_tree_diameter.c
// A program to find the diameter of a binary tree (the longest path between any two nodes)

#include <stdio.h>
#include <stdlib.h>

struct TreeNode {
    int data;
    struct TreeNode* left;
    struct TreeNode* right;
};

int max(int a, int b) {
    return (a > b) ? a : b;
}

int getHeight(struct TreeNode* root) {
    if (root == NULL)
        return 0;
    int leftHeight = getHeight(root->left);
    int rightHeight = getHeight(root->right);
    return 1 + max(leftHeight, rightHeight);
}

int getDiameter(struct TreeNode* root) {
    if (root == NULL)
        return 0;
    int leftHeight = getHeight(root->left);
    int rightHeight = getHeight(root->right);
    int leftDiameter = getDiameter(root->left);
    int rightDiameter = getDiameter(root->right);
    return max(leftHeight + rightHeight + 1, max(leftDiameter, rightDiameter));
}

int main() {
    // Create the binary tree
    struct TreeNode* root = (struct TreeNode*)malloc(sizeof(struct TreeNode));
    root->data = 1;
    root->left = (struct TreeNode*)malloc(sizeof(struct TreeNode));
    root->left->data = 2;
    root->left->left = NULL;
    root->left->right = NULL;
    root->right = (struct TreeNode*)malloc(sizeof(struct TreeNode));
    root->right->data = 3;
    root->right->left = NULL;
    root->right->right = NULL;

    int diameter = getDiameter(root);
    printf("Diameter of the binary tree: %d\n", diameter);
    printf("\n");

    // Clean up memory
    free(root->left);
    free(root->right);
    free(root);

    return 0;
}

File: Data_Structures/Trees/binary_tree_diameter.py
# A program to find the diameter of a binary tree (the longest path between any two nodes)

class TreeNode:
    def __init__(self, data):
        self.data = data
        self.left = None
        self.right = None

def max(a, b):
    return a if a > b else b

def get_height(root):
    if root is None:
        return 0
    left_height = get_height(root.left)
    right_height = get_height(root.right)
    return 1 + max(left_height, right_height)

def get_diameter(root):
    if root is None:
        return 0
    left_height = get_height(root.left)
    right_height = get_height(root.right)
    left_diameter = get_diameter(root.left)
    right_diameter = get_diameter(root.right)
    return max(left_height + right_height + 1, max(left_diameter, right_diameter))

# Create the binary tree
root = TreeNode(1)
root.left = TreeNode(2)
root.right = TreeNode(3)

diameter = get_diameter(root)
print("Diameter of the binary tree:", diameter)
